design-documents/graph-ql/coverage/catalog/configurable-options-selection.md

```diff
@@ -0,0 +1,14 @@
+## Use cases
+
+### Render configurable option values available for selection on the product page
+
+User navigates to the configurable product page. Option values available for selection are rendered on the page.
+The user makes a selection for the first option and the list of option values available for selection is updated for the remaining options. The image relevant for the selection is also updated.
+
+### Add to cart
+
+After the user makes final selection, the corresponding simple product data becomes available and the product can now be added to cart.
+
+### Render configurable option values available for selection on the category page
```

Comment:
> Category and search result pages do not require the list of configurable option values available for selection

I _think_ this might be an unsafe assumption. In a headless world we want to avoid limiting use-cases, especially ones that would allow a shopper to get a product into the cart faster.

Is the idea to not support that at all, or that folks would have to go back to using `variants` field for options when rendering a category?

Reply:
It is not possible to provide selections in category/search queries, unless I am missing something.
 
This is mainly the reason why I decided to have a dedicated query for option selections instead of making them a field of configurable product. The only way to use them would be with empty selections, which does not make a lot of sense because we already have access to full set of option values (not only those available for selection, but all of them).

Reply:
> It is not possible to provide selections in category/search queries, unless I am missing something.

Not quite talking about including it in the query used for rendering the results page. Here is the general use-case I'm talking about:

1. Shopper goes to store, and clicks on "Jeans" in the nav menu.
2. Category page for "Jeans" renders, with product tiles representing each product
3. Each product tile renders some color options (Raw/Acid Wash/Light Wash) and size options (Small/Medium/Large). These are rendered as button elements
4. Shopper clicks "Acid Wash" on the product tile - that product tile should be able to fetch and limit the "Size" options now

As a shopper this is an experience familiar to me on ecommerce websites, and as a front-end dev it's an experience I'd like the API to make possible.

----

> It is not possible to provide selections in category/search queries, unless I am missing something.

Do we expect this to always be true? I'd assume that a search service would support faceted search (cc @kokoc), and I believe facets are driven by attributes? If a search filtered to only include the "Color" attribute with value "Green," I think the client should ideally be able to pre-select that and get back the other options after color selection (like available sizes).

Reply:
On the category use case agree, we can clarify that your use case is supported by the new query. What I meant is that product listing query itself should not support fetching available selections, because all possible selections can be displayed there.

Regarding faceted search, these information may be available in the facet section. Again I don't see a use case why and how we can implement it on the level of the specific product returned in search results. To me this is the same use case as for category and I would not separate them.

Reply:
> Again I don't see a use case why and how we can implement it on the level of the specific product returned in search results. To me this is the same use case as for category and I would not separate them.

It's likely I'm explaining the facet filtering poorly, but I do think this is a separate use-case. Let me try and expand more.

![image](https://user-images.githubusercontent.com/5233399/86984511-68547300-c154-11ea-819a-47a4e9d5f49f.png)

Imagine that this screenshot of Luma is from a headless store built entirely against GraphQL. 

I'm a shopper and I open up the filters on the left and choose "Red." This will limit my results to shorts that come in red, **but** the result set will still have other colors, too.

At this point, the UI knows that I'm trying to drill down to get a red product in my cart. A sufficiently smart UI should now be able to limit the sizes displayed on each tile to only those available in red (because red should be selected).


Keeping the query attached to the `ConfigurableProduct` type enables this flow. Imagine this example query:

```graphql
query FacetedCategoryResults($categoryID: ID, $selectedOptionIDs: [ID!]!) {
   categories(filters: { ids: [$categoryID] }) {
      items {
		...on ConfigurableProduct {
           # Field available on ConfigurableProduct type allows us to do this in 1 query
           options_values(selectedOptionIDs: $selectedOptionIDs) {
              # pick out fields needed for UI
           }
        }
      }
   }
}
```

If we make available options for a selection only available as a root query, the client would have to do N+1 requests to enable this same flow.

Reply:
Thank you for explanation, now your use case is clear. 

In this case we will have to think how it will work in other scenarios involving `ConfigurableProduct` type, like search and other product listing queries. The question there is how should it work when products in the result belong to different attribute sets and thus have different configurable options.

@nrkapoor do we want to go this route to support additional use cases like Andrew described above?

Reply:
Also, have to give credit for the use-case to @soumya-ashok. She had this idea for `pwa-studio` early on and I thought it was pretty cool!

Reply:
@DrewML Thanks for remembering and surfacing the idea we'd discussed. Additionally, if the listing is clicked on to go the PDP, the applied color filter could be the automatic selection.

The same should apply for size or any other filtered parameter.

This should be the more common pattern in Commerce experiences, but surprisingly isn't!

Reply:
Addressed